pandas-stubs/core/dtypes/missing.pyi

```diff
@@ -28,9 +30,17 @@ def isna(obj: Series) -> Series[bool]: ...
 @overload
 def isna(obj: Index | list | ArrayLike) -> npt.NDArray[np.bool_]: ...
 @overload
-def isna(obj: Scalar) -> bool: ...
-@overload
 def isna(obj: NaTType | NAType | None) -> Literal[True]: ...
+@overload
+def isna(
+    obj: Scalar | NaTType | NAType | None,
+) -> TypeGuard[NaTType | NAType | None]: ...
+@overload
+def isna(obj: Scalar | NaTType) -> TypeGuard[NaTType]: ...
+@overload
+def isna(obj: Scalar | NAType) -> TypeGuard[NAType]: ...
+@overload
+def isna(obj: Scalar | None) -> TypeGuard[None]: ...
```

Comment:
These are overlapping overloads. But what's the right way to do this? If you pass a `ScalarT | None`, the type guard should only ever give you a `None`, it should never give you a `NAType` back.

If I remove the last 3 overloads, then you always get back a `TypeGuard[NaTType | NAType | None]`....which isn't correct

Reply:
I think, (but not sure), based on my reading of the docs, that you don't need all the overloads.  If you do 
```python
def isna(
    obj:  NaTType | NAType | None,
) -> TypeGuard[NaTType | NAType | None]: ...
```
that tells the type checker that `isna()` returns `True` for just those types, and `False` for anything else.

What's not clear to me is how to use (and whether we can use) `assert_type()` here.  Functions annotated as `TypeGuard` always return `bool`, but what will `assert_type()` do, both for the type checker and at runtime?  

Keep experimenting.

Reply:
> What's not clear to me is how to use (and whether we can use) `assert_type()` here. Functions annotated as `TypeGuard` always return `bool`, but what will `assert_type()` do, both for the type checker and at runtime?

Could probably have a test like

```py
if not pd.isna(None):
    some_terrible_type_error  # type: ignore[some-unused-code-error] # but not the terrible error which should not be analyzed by mypy/pyright
```

Reply:
I did the following simple test:
```python
from typing_extensions import assert_type, TypeGuard


def isastr(s: str) -> TypeGuard[str]:
    return isinstance(s, str)


assert_type(isastr("abc"), bool)

assert_type(isastr(3.4), bool)

print(assert_type(isastr("asdf"), bool))
```
Both `pyright` and `mypy` indicate that `isastr(3.4)` is invalid.

At runtime, `assert_type(isastr("asdf"), bool)` returns `True`

So for testing purposes, we can use `check(assert_type(pd.isna(foo), bool), True)` where `foo` are `None`, `pd.NA`, `pd.NaT`, `np.nan`.  

Then could do something like `check(assert_type(pd.isna("abc"), bool), False) # type: ignore[arg-type]` to check that we are catching invalid values.

Reply:
To start, the part I'm not following is why we are trying to check the type of the return of the function itself. The new tests I have added to this PR don't bother doing this because that's more or less already covered by existing tests. Instead, they check the type of the object that was passed to `pd.isna`. Because *that* is the object whose type should be influenced by the `TypeGuard`.

The limiting factor I see right now is that yes, `pd.isna` has a variable return type, and `TypeGuard` always needs to return `bool`. But this seems to me a glaring gap in the capability of the type checker - we should be able to narrow types.

If I have this code, which is essentially the same as the test I added:

```
x = random.choice("value", None)  # Optional[str]
if pd.notnull(x):
     ...  # x is guaranteed to be a str
```

We know for a fact that x is a `str` in the body of the `if` statement. I am not aware of any exceptions to that. Anyway, rather than giving up here due to type checking limitations, I'm wondering if this is something worth escalating to the typing mailing list. Is the restriction that `TypeGuard` always returns a `bool` actually necessary? Or can we support `@overload` with `TypeGuard`?

The second, somewhat separate issue, is how the narrowing works. I have a test case illustrating this as well. But to copy the example, this first block of code already works as intended and passes MyPy at least:

```
x = random.choice("value", None, pd.NA)  # Union[str, None, pd.NA]
if pd.isnull(x):
    ...  # x is Union[None, pd.NA]
```

The above code works because we annotate all of the null-equivalent types in a single `Union`. However, if you do that, then this example does NOT work. It should work. But it doesn't:

```
x = random.choice("value", None)  # Union[str, None]
if pd.isnull(x):
    ...  # x should be None, but because we had a single overload that covered pd.NA as well, this is actually Union[None, pd.NA]
```

Now the question I was asking myself, so is the solution to just have separate overloads for each? But you can't, because that overlaps with the original broader overload, and *pyright* complains about overlapping overloads.

So that's essentially where I left off.

Reply:
> To start, the part I'm not following is why we are trying to check the type of the return of the function itself.

This is the methodology we've come up with to test the stubs.  Especially in the case of overloaded functions/methods, we want to make sure that we are getting the expected result for each possible input.

> The limiting factor I see right now is that yes, `pd.isna` has a variable return type, and `TypeGuard` always needs to return `bool`. But this seems to me a glaring gap in the capability of the type checker - we should be able to narrow types.

Is it the capability of the type checker, or the capability of the spec for `TypeGuard` ?

> We know for a fact that x is a `str` in the body of the `if` statement. I am not aware of any exceptions to that. Anyway, rather than giving up here due to type checking limitations, I'm wondering if this is something worth escalating to the typing mailing list. Is the restriction that `TypeGuard` always returns a `bool` actually necessary? Or can we support `@overload` with `TypeGuard`?

Yes, worth bringing up with the typing gods.

Reply:
> This is the methodology we've come up with to test the stubs. Especially in the case of overloaded functions/methods, we want to make sure that we are getting the expected result for each possible input.

I don't think that methodology is by itself sufficient for testing a `TypeGuard`. `TypeGuard` changes the type of the object passed to it - and that's what I've been trying to test here. It's arguably the most relevant aspect to test, otherwise a `TypeGuard` is no better than a `bool` return.  The existing tests, which are associated with the previous `bool` return, still work. To the extent that the previous tests were incomplete I can fill in the gaps. Although at this point seems like this whole thing is on hold since `TypeGuard` in its current state doesn't seem to work here.

> Is it the capability of the type checker, or the capability of the spec for TypeGuard ?

Seems like the spec to me.

Reply:
> Yes, worth bringing up with the typing gods.

I'm not sure how quick the typing mailing list is, but the [discussion page of pyright](https://github.com/microsoft/pyright/discussions) might be a first start where you get very quickly feedback.